docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,348 @@
+```
+
+Dynamic privileges are different from static privileges in the following ways:
+
+* The name of the privilege is “DYNAMIC”. A plugin registers it, and the server has no prior knowledge of its existence.
+* Privileges can only be global scoped.
+* Privileges are stored in the table `mysql.global_grants` and not `mysql.user`.
+* The `GRANT OPTION` is stored for each dynamic privilege (versus for the user as a whole).
+
+We have the same requirement for fine grained privileges in TiDB, so it makes sense to adopt a similar implementation of dynamic privileges. This document describes both the implementation of the framework for dynamic privileges and an initial set of dynamic privileges that are required to be implemented.
+
+## Detailed Design
+
+Implementing Dynamic Privileges requires the following work to be completed.
+
+### Persistence
+
+For TiDB, we can use the same table structure as MySQL:
+
+```sql
+CREATE TABLE `global_grants` (
+  `USER` char(32) NOT NULL DEFAULT '',
+  `HOST` char(255) NOT NULL DEFAULT '',
+  `PRIV` char(32) NOT NULL DEFAULT '',
+  `WITH_GRANT_OPTION` enum('N','Y') NOT NULL DEFAULT 'N',
+  PRIMARY KEY (`USER`,`HOST`,`PRIV`)
+);
+```
+
+There is an existing table called “global_priv” which initially looked like it provided similar functionality, except:
+* The priv is expected to be a JSON encoded string
+* There is no column named `WITH_GRANT_OPTION`.
+
+I looked at repurposing this table (which stores TLS options), but because the `PRIV` value is the data and not the key, it gets messy. I instead plan to use the same schema as MySQL.
+
+This table will persist dynamic privileges. Similar to MySQL, the cache is read into memory and cached (privilege/privileges/cache.go). Dynamic privileges will be cached in the same way as existing privileges.
+
+### Privilege Checking API
+
+Checking for existence of a Dynamic privilege needs a different function from normal privilege checks. I.e.
+
+```
+// RequestVerification(activeRole []*auth.RoleIdentity, db, table,
+// column string, priv mysql.PrivilegeType) bool
+if pm.RequestVerification(activeRoles, "", "", "", mysql.ProcessPriv) {
+	// has processPrivilege
+}
+```
+
+This is not suitable because:
+* The privilege `mysql.ProcessPriv` must be predefined (i.e. it's not dynamic).
+* The 3 empty string values (schema, table, column) are never applicable to dynamic privileges.
+* Dynamic privileges are grantable individually. There may be scenarios where code wants to check if a user has both a dynamic privilege and the ability to grant it (such as in the output of `SHOW GRANTS`).
+
+I propose the following:
+
+```
+// RequestDynamicVerification(activeRole []*auth.RoleIdentity, priv string, grantable bool) bool
+if pm.RequestDynamicVerification(activeRoles, "BACKUP_ADMIN", false) {
+	// has backup admin privilege
+}
+```
+
+### Plugin API
+
+There will need to be a way for plugins to register new dynamic privileges via their OnInit method. I propose the following:
```

Comment:
The plugin registers the permission, but the binding is loose. If the plugin is uninstalled there could be users who still have grants assigned to those dynamic names. I checked the intended behavior in MySQL, which is to report error `ErrDynamicPrivilegeNotRegistered = 3929` in revoke context, and parse error in `GRANT` context.

My plan is to implement a technically incompatble (but improved behavior) or returning 3929 in both instances. For example see: https://github.com/pingcap/tidb/pull/22778/files#diff-83d52377771c642f1bbe6ecaf9a98e6d3787f57c715fb468e26abce2f3d51dadR437-R442 and https://github.com/pingcap/tidb/pull/22778/files#diff-fc0b54ba2dcfe653ca33434b6eabd9d5a06e87883827b29deddc134f6941fb05R193-R195

In MySQL, uninstalling a plugin is only effective upon server restart. I believe the same is true of TiDB plugins, but effectively because there are multiple TiDB instances we need to expect that some servers may effectively not have all plugins loaded.

An alternative design would be to permanently retain registered `DYNAMIC` privileges in an internal table. The upside of this is fewer errors to clients. The downside is that it can produce a similar semantic to "parse but ignore", because it accepts input for which the plugins are not correctly loaded and will never be able to correctly use this functionality. i.e. it is a feature to return an error, not a bug.